### ISSUE TITLE
OAuth2 invalid redirection to localhost:3200 (SwaggerUi3)

### ISSUE THREAD

**Issue by zuckerthoben** [RSuter/NSwag#961]
When you switch from .UseSwaggerUi() to .UseSwaggerUi3() and try to use OAuth 2.0 Authorization in the Swagger Ui, the redirection after signing in (to e.g. Azure AD) always wants to go to localhost:3000. 

So: 
```
app.UseSwaggerUi(typeof(Startup).GetTypeInfo().Assembly, new SwaggerUiSettings()
    {
        OAuth2Client = new OAuth2ClientSettings
        {
            ClientId = clientId,
            ClientSecret = clientSecret,
            AppName = "App",
                Realm = realm,
                AdditionalQueryStringParameters =
                {
                    { "resource", clientId }
                }
           },
          DocumentProcessors =
          {
              new SecurityDefinitionAppender("oauth2", new SwaggerSecurityScheme
              {
                   Type = SwaggerSecuritySchemeType.OAuth2,
                   Description = "DocCheck",
                   Flow = SwaggerOAuth2Flow.Implicit,
                   AuthorizationUrl = authorizationUrl,
                   TokenUrl = tokenUrl,
               })
           },
           OperationProcessors =
           {
               new OperationSecurityScopeProcessor("oauth2")
           },
           DefaultPropertyNameHandling = PropertyNameHandling.CamelCase,
    });
```

works correctly.  
If you change it to SwaggerUi3 and SwaggerUi3Settings without changing anything else, you always get redirected to localhost:3000 after signing in. I guess its hardcoded somewhere?

**Comment by RSuter** [comment]
Have you checked the served HTML of Swagger UI 3? The auth settings are replaced here: 

https://github.com/RSuter/NSwag/blob/master/src/NSwag.AspNetCore/SwaggerUi3/index.html#L99

Maybe we need more settings to make this work?

**Comment by RSuter** [comment]
Can you try to host these files yourself (in your project):  

https://github.com/RSuter/NSwag/tree/master/src/NSwag.AspNetCore/SwaggerUi3

and find a solution to this problem? 

Replace index.html with the served index.html and modify it so that your scenario works...

As soon as we know how to fix this, we can implement this in NSwag...

**Comment by pekkah** [comment]
Hitting this issue. ASP.NET Core apps are hosted behind a reverse proxy with IIS. Somehow the internal url is getting used as the redirect url for the oauth instead of the IIS specific address. Easiest solution would be to allow giving the redirect url in the C# OAuth2ClientSettings and using that when initializing the JS client.

**Comment by pekkah** [comment]
Upon close inspection the wrong url was just a false positive on the identity server logs. The actual issue seems to be that the patch for the redirect is wrong. If my app is hosted at https://localhost/api/app then the redirect url is sent as https://localhost/oauth2-redirect.html. Is there somewhere option to set the base path?

**Comment by pekkah** [comment]
After deepdive it seems that the problem is not with any reverse proxy thingy either.
It seems that the url is coming from here https://github.com/swagger-api/swagger-ui/blob/f22a6289345a3025650c687ab482a9e1a6369b07/swagger-config.yaml. How do I change it with the middlware?

**Comment by RSuter** [comment]
Currently there are not many options in the Swagger UI 3 middleware... We need to add the missing options.

**Comment by softarch74** [comment]
Any ETA?

**Comment by RSuter** [comment]
Added the setting ServerUrl => check my commit. Is this how this should work?

**Comment by softarch74** [comment]
Yes, that is my understanding of how it's suppose to work.

**Comment by stijnherreman** [comment]
I'm not sure if the new `ServerUrl` property is required for anyone, but the default (empty setting) should make it use the current URL. Now you get `redirect_uri=%2Fswagger%2Foauth2-redirect.html`, which cannot work afaik, it must be an absolute URL.

**Comment by RSuter** [comment]
If you use Oauth2/Redirect you also need to set the ServerUrl property... isnt it working if it is correctly set?

**Comment by stijnherreman** [comment]
That requires knowing the URL beforehand. It's possible, but it's not very flexible, and I'd prefer not introducing such configuration in my code.

**Comment by RSuter** [comment]
So, should we use `window.location.origin` when ServerUrl is not set?

**Comment by stijnherreman** [comment]
That would cause problems when the site is configured as an application in IIS (e.g. `http://myserver/MySite`). The safest way would be to go from the current URL `https://localhost:44321/swagger/index.html?url=/swagger/v1/swagger.json` to `https://localhost:44321/swagger/oauth2-redirect.html`, but I'm not sure what the easiest JS method is to do that.

**Comment by RSuter** [comment]
Do you think this would not work?

```
            html = html.Replace("{DocExpansion}", DocExpansion);
            html = html.Replace("{RedirectUrl}", string.IsNullOrEmpty(ServerUrl) ?
                "window.location.origin + \"" + SwaggerUiRoute + "/oauth2-redirect.html\"" :
                "\"" + ServerUrl + SwaggerUiRoute + "/oauth2-redirect.html\"");
```

**Comment by stijnherreman** [comment]
Perhaps, it depends on what `SwaggerUiRoute` resolves to. If it handles a site being hosted at a location like `http://myserver/MySite/swagger/index.html`, then yes it would work.

**Comment by RSuter** [comment]
I think in this case, you would set 

MiddlewareBasePath = /MySite
SwaggerUiRoute = /swagger

and it should work...

**Comment by stijnherreman** [comment]
Thanks, I'll give it a try tomorrow.

**Comment by RSuter** [comment]
Ok, but this change is not released yet...

**Comment by Aveline67** [comment]
I can reproduce this issue using Swashbuckle.AspNetCore.SwaggerUI 2.2.0 on AspNetCore app

**Comment by RSuter** [comment]
Can we close this?

**Comment by softarch74** [comment]
I have moved away from using it, so I couldn't say. Sorry.

John Morris
Sr. Software Architect
SoftPro
________________________________

**Comment by jfidot** [comment]
I had the same issue using NSwag 11.18.6 on .NET Framework 4.6.1 web api. 

The solution was set the Setting:

`SwaggerUi3Settings.ServerUrl = "..."`

@RSuter your fix from 19 Dec 2017 seems to work properly. Thanks!

**Comment by Rutgerz** [comment]
The earlier reference to MiddlewareBasePath did not work for me. However the ServerUrl property does work.  I know use this to correctly configure the url. 
`settings.ServerUrl = env.IsDevelopment() ? "https://localhost:44399/" : "https://api.domain.com/v1";`

**Comment by jmatter** [comment]
I still have this issue using NSwag.AspNetCore 11.20.1.
I set the API Base URL to `/api`:
`app.UsePathBase("/api");`

I haven't set `ServerUrl`, because the API is published to multiple environments and I don't want to deal with absoute URLs in the code.

So the URL the API runs is `https://domain.com/api/`
Swagger is reachable on `https://domain.com/api/swagger`.

The generated oauth2 redirect URL is `https://domain.com/swagger/oauth2-redirect.html` but should be `https://domain.com/api/swagger/oauth2-redirect.html`.

So it seems the `app.UsePathBase("/api");` is not incorporated into `SwaggerUiRoute`.

The suggested solution by @stijnherreman 
> 
> ```js
> window.location.origin + window.location.pathname.replace("index.html", "oauth2-redirect.html")
> ```
would work in this scenario and I think it should work in every case...

**Comment by OculiViridi** [comment]
I've a similar problem.
On **Development environment**, the application runs on Kestrel.
Application runs on: `http://localhost:5001`.
Swagger runs on: `http://localhost:5001/swagger`.

On **Production environment**, the application runs on Kestrel _behind a Reverse Proxy_.
The domain is: `dev.mydomain.com`
The application has to run locally on the dev server on: `http://myservername:5001`
The application has to be reachable from outside on: `http://dev.mydomain.com/control` (thanks to the reverse proxy)
Swagger has to be reachable from outside on: `http://dev.mydomain.com/control/swagger`

So, there's the need to bind `http://myservername:5001` to `http://dev.mydomain.com/control`. But actually the reverse proxy settings is not enough.

What are all the necessary settings and where are located?
Obviously, I want to avoid hard-coded strings and paths, because the application should be able to be moved to another server/environment without needing a recompilation.

**Comment by RSuter** [comment]
Same problem here: https://github.com/RSuter/NSwag/issues/1717

**Comment by RSuter** [comment]
I've created a new PR with a new setting: https://github.com/RSuter/NSwag/pull/1728
See #1717 (bottom) for more information

**Comment by RSuter** [comment]
Closing issue as original problem is solved and discussions are now in this issue: #1717
